definitions/mw.lua

```diff
@@ -0,0 +1,649 @@
+
+---Returns an object with one method, object:expand(), that returns the result of frame:preprocess( text ).
+---@param params {text: string}
+---@return {expand: fun(self):string}
+---@overload fun(self, text: string): {expand: fun(self):string}
+function mw.frame:newParserValue(params) end
+
+---Returns an object with one method, object:expand(), that returns the result of frame:expandTemplate called with the given arguments.
+---@param params {title: string, args: table}
+---@return {expand: fun(self):string}
+function mw.frame:newTemplateParserValue(params) end
+
+mw.hash = {}
+
+---Hashes a string value with the specified algorithm. Valid algorithms may be fetched using mw.hash.listAlgorithms().
+---@param algo string
+---@param value any
+---@return string
+function mw.hash.hashValue(algo, value) end
+
+---Hashes a string value with the specified algorithm. Valid algorithms may be fetched using mw.hash.listAlgorithms().
+---@return string[]
+function mw.hash.listAlgorithms() end
+
+---@class Html
+mw.html = {}
+
+---Creates a new mw.html object containing a tagName html element. You can also pass an empty string or nil as tagName in order to create an empty mw.html object.
+---@param tagName? string
+---@param args? {selfClosing: boolean, parent: Html}
+---@return Html
+function mw.html.create(tagName, args) end
+
+---Appends a child mw.html (builder) node to the current mw.html instance. If a nil parameter is passed, this is a no-op. A (builder) node is a string representation of an html element.
+---@param builder? Html|string
+---@return self
+function mw.html:node(builder) end
+
+---Appends an undetermined number of wikitext strings to the mw.html object. Note that this stops at the first nil item.
+---@param ... string?
+---@return self
+function mw.html:wikitext(...) end
+
+---Appends a newline to the mw.html object. 
+---@return self
+function mw.html:newline() end
+
+---Appends a new child node with the given tagName to the builder, and returns a mw.html instance representing that new node. The args parameter is identical to that of mw.html.create
+---Note that contrarily to other methods such as html:node(), this method doesn't return the current mw.html instance, but the mw.html instance of the newly inserted tag. Make sure to use html:done() to go up to the parent mw.html instance, or html:allDone() if you have nested tags on several levels.
+---@param tagName string
+---@param args? {selfClosing: boolean, parent: Html}
+---@return Html
+function mw.html:tag(tagName, args) end
+
+---Set an HTML attribute with the given name and value on the node. Alternatively a table holding name->value pairs of attributes to set can be passed. In the first form, a value of nil causes any attribute with the given name to be unset if it was previously set.
+---@param name string
+---@param value string
+---@return self
+---@overload fun(self, param: {[string]: string})
+function mw.html:attr(name, value) end
```

Comment:
this only accounts for one of the allowed input patterns
`Alternatively a table holding name->value pairs of attributes to set can be passed.`

Reply:
The overload has the second variant